File: experiments/adrianco/experiment-5/runs/language=python_model=claude-opus-4-8_tooling=beads/rep1-failed/brazilian_soccer/queries/competitions.py
"""
================================================================================
Context
================================================================================
Project   : Brazilian Soccer MCP Server
Module    : brazilian_soccer.queries.competitions
Purpose   : Competition-level views, primarily league standings calculated from
            match results (3 points for a win, 1 for a draw).

Functions
  standings(kg, competition, season)  -> sorted league table with points, W/D/L,
                                          goals for/against, goal difference.
  champion(kg, competition, season)   -> the top team of the computed standings.
  relegated(kg, season, count)        -> bottom `count` teams of a Brasileirao
                                          season standings.

Standings are most meaningful for round-robin leagues (Brasileirao); they can be
computed for any competition but a knockout cup table is just a points summary.
================================================================================
"""

from __future__ import annotations

from typing import List, Optional

from ..knowledge_graph import KnowledgeGraph


def standings(
    kg: KnowledgeGraph,
    competition: str = "Brasileirao",
    season: Optional[int] = None,
) -> List[dict]:
    """Compute a league table for a competition/season from match results."""
    table: dict[str, dict] = {}

    for m in kg.matches_by_competition.get(competition, []):
        if season is not None and m.season != season:
            continue
        if not m.has_score:
            continue

        for key, gf, ga in (
            (m.home_key, m.home_goal, m.away_goal),
            (m.away_key, m.away_goal, m.home_goal),
        ):
            row = table.setdefault(
                key,
                {
                    "team": kg.display_name(key),
                    "played": 0,
                    "wins": 0,
                    "draws": 0,
                    "losses": 0,
                    "goals_for": 0,
                    "goals_against": 0,
                    "points": 0,
                },
            )
            row["played"] += 1
            row["goals_for"] += gf
            row["goals_against"] += ga
            if gf > ga:
                row["wins"] += 1
                row["points"] += 3
            elif gf < ga:
                row["losses"] += 1
            else:
                row["draws"] += 1
                row["points"] += 1

    rows = list(table.values())
    for r in rows:
        r["goal_difference"] = r["goals_for"] - r["goals_against"]

    rows.sort(
        key=lambda r: (r["points"], r["goal_difference"], r["goals_for"]),
        reverse=True,
    )
    for i, r in enumerate(rows, start=1):
        r["position"] = i
    return rows


def champion(
    kg: KnowledgeGraph,
    competition: str = "Brasileirao",
    season: Optional[int] = None,
) -> Optional[dict]:
    """Return the top team of the computed standings, or None if no data."""
    table = standings(kg, competition, season)
    return table[0] if table else None


def relegated(
    kg: KnowledgeGraph,
    season: int,
    competition: str = "Brasileirao",
    count: int = 4,
) -> List[dict]:
    """Return the bottom `count` teams of a season's standings."""
    table = standings(kg, competition, season)
    return table[-count:] if table else []

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-8_tooling=beads/rep1/src/queries/competitions.ts
/**
 * ============================================================================
 * Context
 * ----------------------------------------------------------------------------
 * Module:  src/queries/competitions.ts
 * Purpose: Competition-level queries — league standings computed from match
 *          results (3 points for a win, 1 for a draw), and the champion /
 *          relegation picture for a season.
 *
 * Backs the MCP tools `standings` and `season_summary`. Standings are computed
 * from deduplicated matches so the three overlapping Serie A sources do not
 * inflate points. Sorting follows the Brazilian convention: points, then wins,
 * then goal difference, then goals for.
 * ============================================================================
 */

import type { Dataset, Match } from "../data/types.js";
import { dedupeMatches, filterMatches } from "./common.js";

export interface StandingRow {
  rank: number;
  team: string;
  played: number;
  wins: number;
  draws: number;
  losses: number;
  goalsFor: number;
  goalsAgainst: number;
  goalDiff: number;
  points: number;
}

export interface StandingsResult {
  competition: string;
  season: number;
  rows: StandingRow[];
  text: string;
}

interface Acc {
  team: string;
  played: number;
  wins: number;
  draws: number;
  losses: number;
  goalsFor: number;
  goalsAgainst: number;
}

/** Compute a league table for a competition + season from match results. */
export function standings(
  ds: Dataset,
  competition: string,
  season: number
): StandingsResult {
  const games = dedupeMatches(
    filterMatches(ds.matches, { competition, season })
  );

  // Group by normalized key (so differing display spellings of the same club
  // merge into one row), keeping the first-seen display name for output.
  const table = new Map<string, Acc>();
  const get = (key: string, display: string): Acc => {
    let a = table.get(key);
    if (!a) {
      a = {
        team: display,
        played: 0,
        wins: 0,
        draws: 0,
        losses: 0,
        goalsFor: 0,
        goalsAgainst: 0,
      };
      table.set(key, a);
    }
    return a;
  };

  for (const m of games) {
    if (m.homeGoal == null || m.awayGoal == null) continue;
    const h = get(m.homeKey, m.homeTeam);
    const a = get(m.awayKey, m.awayTeam);
    h.played++;
    a.played++;
    h.goalsFor += m.homeGoal;
    h.goalsAgainst += m.awayGoal;
    a.goalsFor += m.awayGoal;
    a.goalsAgainst += m.homeGoal;
    if (m.homeGoal > m.awayGoal) {
      h.wins++;
      a.losses++;
    } else if (m.homeGoal < m.awayGoal) {
      a.wins++;
      h.losses++;
    } else {
      h.draws++;
      a.draws++;
    }
  }

  const rows: StandingRow[] = [...table.values()]
    .map((a) => ({
      rank: 0,
      team: a.team,
      played: a.played,
      wins: a.wins,
      draws: a.draws,
      losses: a.losses,
      goalsFor: a.goalsFor,
      goalsAgainst: a.goalsAgainst,
      goalDiff: a.goalsFor - a.goalsAgainst,
      points: a.wins * 3 + a.draws,
    }))
    .sort(
      (x, y) =>
        y.points - x.points ||
        y.wins - x.wins ||
        y.goalDiff - x.goalDiff ||
        y.goalsFor - x.goalsFor ||
        x.team.localeCompare(y.team)
    );
  rows.forEach((r, i) => (r.rank = i + 1));

  const lines = rows.map(
    (r) =>
      `${String(r.rank).padStart(2)}. ${r.team} — ${r.points} pts ` +
      `(${r.wins}W ${r.draws}D ${r.losses}L, GF ${r.goalsFor} GA ${r.goalsAgainst}, GD ${r.goalDiff >= 0 ? "+" : ""}${r.goalDiff})`
  );
  const header =
    rows.length > 0
      ? `${competition} ${season} — final standings (calculated from ${games.length} matches):`
      : `No matches found for ${competition} ${season}.`;
  const text = `${header}${lines.length ? "\n" + lines.join("\n") : ""}`;

  return { competition, season, rows, text };
}

export interface SeasonSummary {
  competition: string;
  season: number;
  champion: string | null;
  relegated: string[];
  totalTeams: number;
  text: string;
}

/**
 * Summarize a league season: champion (top of table) and the relegation zone
 * (bottom 4, the Brazilian Série A convention) when there are enough teams.
 */
export function seasonSummary(
  ds: Dataset,
  competition: string,
  season: number,
  relegationSpots = 4
): SeasonSummary {
  const { rows } = standings(ds, competition, season);
  const champion = rows[0]?.team ?? null;
  const relegated =
    rows.length > relegationSpots
      ? rows.slice(rows.length - relegationSpots).map((r) => r.team)
      : [];

  const text =
    rows.length === 0
      ? `No data for ${competition} ${season}.`
      : `${competition} ${season}: champion ${champion} (${rows[0].points} pts).` +
        (relegated.length
          ? `\nRelegation zone (bottom ${relegationSpots}): ${relegated.join(", ")}`
          : "");

  return {
    competition,
    season,
    champion,
    relegated,
    totalTeams: rows.length,
    text,
  };
}
